### PR TITLE
[DmaComposition] Rework logic to enable new composition opportunity

### PR DESCRIPTION
**Pull request by newling**
- slight modification to logic used for compositing DMAs, makes use of global offset difference (please see comments in code)
- more aggressive folding of unit dimensions 
- new unit test in combine_strided_ops.mlir shows the case that was being missed in an end-to-end matmul transpose b with the new pipeline pack-peel-4-level-tiling

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp [hunk 1]
```diff
@@ -511,6 +480,91 @@ LogicalResult moveNpuDmaSyncUsersAfterAncestorInSameBlock(
   return success();
 }
 
+/// Try to add `offsetToMerge` to one of the offsets in `offsets`.  This
+/// function assumes an effective stride for `offsetToMerge` of 1one so to add
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:484]: I'm not sure what it should be. 1one -> one?

**Review by yzhang93 — COMMENTED** [review]
LGTM. Some minor comments.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp (hunk 1 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:484]: eleven ;) 
I'll fix, thanks.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp [hunk 2]
```diff
@@ -326,6 +332,9 @@ TEST_F(AccessPatternCombinationTest, CombineAccessPatterns) {
   checkCombineAccessPatterns({0, 0}, {16, 32}, {16, 1}, {1, 0, 0}, {3, 16, 32},
                              {64, 16, 1}, {0, 0, 0}, {4, 16, 32}, {64, 16, 1},
                              3);
+  // stride matching
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp:335]: Could you add a few more test cases for this case with more dimensions and with non-zero offsets for A? E.g. something like this:
```
checkCombineAccessPatterns({4, 0}, {8, 16}, {64, 1}, {4 * 64 + 8 * 64, 0}, {1, 16},
                             {1, 1}, {4, 0}, {9, 16}, {64, 1}, 3, true);
```

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp [hunk 3]
```diff
@@ -7,201 +7,275 @@
...
+///
+/// This function computes the difference between the global offsets of two
+/// access patterns. If it is not constant, i.e. if the difference contains
+/// an MLIR value which is not a constant, then nullopt is returned.
+///
+/// This function is useful when determining if the access pattern A, followed
+/// by the access pattern B, can be merged into a single access pattern.
+///
+/// \return global_offset(X) - global_offset(Y).
+std::optional<int64_t> getGlobalOffsetDifference(
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:100]: I think it would be useful to add unit tests for this utility (and other similar utilities).

**Review by jtuyls — CHANGES_REQUESTED** [review]
This is a great refactor with additional combinations being enabled, thanks for the thorough logic and great explanations!

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp [hunk 4]
```diff
@@ -7,201 +7,275 @@
...
+      globalOffset += sign * o.value() * s.value();
+    } else if (o.has_value()) {
+      increment(cast<Value>(stride), sign * o.value());
+    } else if (s.has_value()) {
+      increment(cast<Value>(offset), sign * s.value());
     }
+    return true;
+  };
+
+  for (uint32_t i = 0; i < offsetsX.size(); ++i) {
```
[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:139]: That was what I had originally, but then there are missed opportunities where the difference is constant, but the individual global offsets are not. I hit these cases with some of the end-to-end matmuls. 

Something like 
X = offsets:  [%v] sizes: [10] strides: [1]
Y = offsets:  [%v]  sizes: [10] strides: [1]

for example: the global offset of X (and Y) is %v  (i.e. non-constant), but the difference is %v - %v = 0 is constant. So by computing the global difference directly we see we can merge to 

merged = offsets: [0, %v] sizes: [2, 10] strides: [0, 1]. 

Another option could be to actually build a global offset expression with arith ops for X and Y, and then hope to somehow canonicalize the difference between them and hope to get a constant. This is what I had originally, but it didn't work.  

I will add more tests with non-constant offsets illustrating this.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp (hunk 2 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp]: I've added this and a few others

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp (hunk 3 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:50]: I've added some in the .cpp test file for the constant cases. For the non-constant cases, I'm in the process of adding in the combine_strided_ops.mlir, need to add some negative cases though (non-constant global offset difference). Tomorrow..

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jtuyls) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/Utils/AMDAIEDmaUtils.cpp:123]: Thanks for the explanation! I think it would be useful to add this to the docstring of the function as well. Can be brief, as a reference for someone touching that function in the future.

**Comment by newling** [comment]
> This is a great refactor with additional combinations being enabled, thanks for the thorough logic and great explanations!

Thanks! I think I've addressed all the comments now, let me know if you think it should have more testing etc.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp [hunk 5]
```diff
@@ -6,15 +6,27 @@
...
 #include "llvm/ADT/SmallVectorExtras.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
 
 namespace {
 
 using namespace mlir;
 using namespace mlir::iree_compiler::AMDAIE;
 
+SmallVector<int64_t> fromOpFoldResults(SmallVector<OpFoldResult> ofrs) {
```
Comment (jtuyls) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp:18]: Can't you use `getConstantIntValues`?

**Review by jtuyls — APPROVED** [review]
LGTM, just one nit above.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp (hunk 5 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/AMDAIEDmaUtilsTest.cpp:18]: Oops second time. Changed
